### PR TITLE
Faster contains_full_disk

### PR DESCRIPTION
**Pull request by wafels**
### Description

This PR speeds up checking that a map contains the full disk.  I get about a ~20x speed up (2016 MacBook Pro, 13'')

This PR is a direct result of the AIApy acceleration project, part of HelioHackweek 2020 - see https://github.com/heliohackweek/AIA_acceleration/issues/2

### CODE REVIEW

File: sunpy/map/maputils.py [hunk 1]
```diff
@@ -124,15 +124,38 @@ def sample_at_coords(smap, coordinates):
...
 
 
+def _convert_pixels(edge):
+    """
+    Helper function to convert a list of edge pixels of the form [(x0, y0), (x1, y1),..., (xn,yn)]
+    into the form ([x0, x1,...,xn], [y0, y1,...,yn])
+
+    Parameter
+    ---------
+    edge : ~list
```
Comment (nabobalis) [inline sunpy/map/maputils.py:134]:
```suggestion
    edge : list
```

File: sunpy/map/maputils.py [hunk 2]
```diff
@@ -124,15 +124,38 @@ def sample_at_coords(smap, coordinates):
...
+        A list of pairs of tuples.
+
+    Returns
+    -------
+    pixel_edge : ~tuple
```
Comment (nabobalis) [inline sunpy/map/maputils.py:139]:
```suggestion
    tuple
```

File: sunpy/map/maputils.py [hunk 3]
```diff
@@ -124,15 +124,38 @@ def sample_at_coords(smap, coordinates):
...
+        A tuple containing two lists.  The first entry is a list containing the first coordinate of
```
Comment (nabobalis) [inline sunpy/map/maputils.py:140]:
```suggestion
        A tuple containing two lists. The first entry is a list containing the first coordinate of
```

File: sunpy/map/maputils.py [hunk 4]
```diff
@@ -124,15 +124,38 @@ def sample_at_coords(smap, coordinates):
...
+        each pixel, and similarly for the second entry.
+
```
Comment (nabobalis) [inline sunpy/map/maputils.py:142]:
```suggestion
```

File: sunpy/map/maputils.py [hunk 5]
```diff
@@ -153,21 +176,66 @@ def contains_full_disk(smap):
...
-
-    # Test if all the edge pixels are more than one solar radius distant
-    # and that the whole map is not all off disk.
-    return np.all(coordinate_angles > solar_angular_radius(edge_of_world)) and ~is_all_off_disk(smap)
+    # Get the edge pixels
+    top_, bottom, left_hand_side, right_hand_side = map_edges(smap)
+
+    # Coordinates of the edges
+    x, y = _convert_pixels(top_)
+    horizontal1 = smap.pixel_to_world(x, y)
```
Comment (nabobalis) [inline sunpy/map/maputils.py:184]:
```suggestion
    horizontal1 = smap.pixel_to_world(_convert_pixels(top_))
```
Is there a better name than `horizontal1`? (The same applies for the other lines.

File: sunpy/map/maputils.py (hunk 1 above)
Comment (wafels) [reply sunpy/map/maputils.py]: Fixed

File: sunpy/map/maputils.py (hunk 2 above)
Comment (wafels) [reply sunpy/map/maputils.py]: Fixed

File: sunpy/map/maputils.py (hunk 3 above)
Comment (wafels) [reply sunpy/map/maputils.py]: Fixed

File: sunpy/map/maputils.py (hunk 4 above)
Comment (wafels) [reply sunpy/map/maputils.py]: Fixed

File: sunpy/map/maputils.py [hunk 6]
```diff
@@ -153,21 +175,59 @@ def contains_full_disk(smap):
...
-    coordinate_angles = np.sqrt(edge_of_world.Tx ** 2 + edge_of_world.Ty ** 2)
-
-    # Test if all the edge pixels are more than one solar radius distant
-    # and that the whole map is not all off disk.
-    return np.all(coordinate_angles > solar_angular_radius(edge_of_world)) and ~is_all_off_disk(smap)
+    # Get the edge pixels
+    top_, bottom, left_hand_side, right_hand_side = map_edges(smap)
+
+    # Coordinates of the edges
+    horizontal1 = smap.pixel_to_world(*_convert_pixels(top_))
```
Comment (nabobalis) [inline sunpy/map/maputils.py:182]: I forgot about the need to unpack it.

File: sunpy/map/maputils.py [hunk 7]
```diff
@@ -153,21 +175,59 @@ def contains_full_disk(smap):
...
+        rhs = v1
+    elif np.all(v1 < -radius):
+        lhs = v1
+
+    v2 = vertical2.Tx
+    if np.all(v2 > radius):
+        rhs = v2
+    elif np.all(v2 < -radius):
+        lhs = v2
+
```
Comment (nabobalis) [inline sunpy/map/maputils.py:224]: These basically repeat but changed out for a different variable. There should be a neater way to write this but I don't have an answer at the moment.

File: sunpy/map/maputils.py (hunk 5 above)
Comment (wafels) [reply sunpy/map/maputils.py]: I struggled coming up with these names.  For a N x N image, the `top_ `pixels are (0, N-1),...(N-1, N-1), that is, the second axis value is constant at its maximum value.  A more descriptive name would be `second_axis_value_is_max`.   Following this scheme, we also have `second_axis_value_is_min`, `first_axis_value_is_max` and `first_axis_value_is_min`.  Is this better?

File: sunpy/map/maputils.py (hunk 6 above)
Comment (wafels) [reply sunpy/map/maputils.py:182]: Fixed.

File: sunpy/map/maputils.py (hunk 7 above)
Comment (wafels) [reply sunpy/map/maputils.py:224]: Yeah, I agree.  I wrote the function quickly and in a long-winded way to support the hackweek.

File: sunpy/map/maputils.py [hunk 8]
```diff
@@ -124,15 +124,37 @@ def sample_at_coords(smap, coordinates):
...
+    Returns
+    -------
+    tuple
+        A tuple containing two lists. The first entry is a list containing the first coordinate of
+        each pixel, and similarly for the second entry.
+    """
+    x = [p[0] for p in edge.value] * u.pix
+    y = [p[1] for p in edge.value] * u.pix
+    return x, y
```
Comment (ayshih) [inline sunpy/map/maputils.py:145]: This looks slow (performance-wise).  At the very least, the list comprehension should not be done twice.

**Comment by ayshih** [comment]
If you want performance, I think you'd do better if you reversed the check (at least for high-resolution images.  Instead of making sure all of the edges of the map are at least a solar radius away, how about creating an array coordinate for the limb of the Sun and seeing if it all lands in the array when converted into pixel space?

**Comment by wafels** [comment]
I'll try that.

File: sunpy/map/maputils.py (hunk 8 above)
Comment (wafels) [reply sunpy/map/maputils.py:145]: I could just replace `map_edges` with a function that returns the pixels in the form that they are needed.

Comment (Cadair) [reply sunpy/map/maputils.py:145]: isn't this `zip(egde.value)`? or the reverse of that maybe?

**Comment by nabobalis** [comment]
If we can benchmark this using https://github.com/joerick/pyinstrument would be nice.

**Comment by nabobalis** [comment]
I think that #4463 has become more advanced and supersedes this.

**Pull request closed without merging**